File: ds9/tools/inspector-result-validator.mjs
#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';

function usage() {
  console.error('Usage: node tools/inspector-result-validator.mjs <path-to-inspector_result.json>');
  process.exit(1);
}

function validateInspectorResult(obj) {
  const errors = [];

  if (!obj || typeof obj !== 'object' || Array.isArray(obj)) {
    errors.push('root must be a JSON object');
    return errors;
  }

  const status = obj.status;
  const allowedStatus = new Set(['approved', 'changes_requested']);
  if (typeof status !== 'string' || !allowedStatus.has(status)) {
    errors.push('status must be "approved" or "changes_requested"');
  }

  const issues = obj.issues;
  if (!Array.isArray(issues)) {
    errors.push('issues must be an array');
  } else {
    const allowedSev = new Set(['blocker', 'major', 'minor']);
    issues.forEach((issue, index) => {
      if (!issue || typeof issue !== 'object' || Array.isArray(issue)) {
        errors.push(`issues[${index}] must be an object`);
        return;
      }
      const i = issue;
      if (typeof i.severity !== 'string' || !allowedSev.has(i.severity)) {
        errors.push(`issues[${index}].severity must be one of: blocker | major | minor`);
      }
      if (typeof i.description !== 'string' || i.description.trim() === '') {
        errors.push(`issues[${index}].description must be a non-empty string`);
      }
      if (!Array.isArray(i.paths)) {
        errors.push(`issues[${index}].paths must be an array of strings`);
      } else {
        i.paths.forEach((p, pi) => {
          if (typeof p !== 'string') {
            errors.push(`issues[${index}].paths[${pi}] must be a string`);
          }
        });
      }
    });
  }

  const nextTasks = obj.next_tasks;
  if (!Array.isArray(nextTasks)) {
    errors.push('next_tasks must be an array of strings');
  } else {
    nextTasks.forEach((t, i) => {
      if (typeof t !== 'string') {
        errors.push(`next_tasks[${i}] must be a string`);
      }
    });
  }

  if (status === 'changes_requested') {
    if (Array.isArray(issues) && issues.length === 0) {
      errors.push('status "changes_requested" requires at least one issue');
    }
    if (Array.isArray(nextTasks) && nextTasks.length === 0) {
      errors.push('status "changes_requested" requires at least one next_tasks entry');
    }
  }

  return errors;
}

function main() {
  const args = process.argv.slice(2);
  if (args.length !== 1) {
    usage();
  }

  const filePath = path.resolve(process.cwd(), args[0]);

  let raw;
  try {
    raw = fs.readFileSync(filePath, 'utf8');
  } catch (err) {
    console.error('inspector-result-validator: failed to read file:', err.message);
    process.exit(1);
  }

  let data;
  try {
    data = JSON.parse(raw);
  } catch (err) {
    console.error('inspector-result-validator: invalid JSON:', err.message);
    process.exit(1);
  }

  const errors = validateInspectorResult(data);
  if (errors.length > 0) {
    console.error('inspector-result-validator: schema validation failed:');
    for (const e of errors) console.error(' -', e);
    process.exit(1);
  }

  console.log('inspector-result-validator: OK');
}

main();

File: .foreman/inspector/tools/inspector_result_validator.py
#!/usr/bin/env python3
import json
import sys
from dataclasses import dataclass
from typing import Any, Dict, List, Optional


@dataclass
class ValidationError:
    path: str
    message: str
    code: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        data: Dict[str, Any] = {"path": self.path, "message": self.message}
        if self.code is not None:
            data["code"] = self.code
        return data


def load_input() -> Dict[str, Any]:
    try:
        raw = sys.stdin.read()
        return json.loads(raw or "{}")
    except json.JSONDecodeError as exc:
        raise SystemExit(json.dumps({
            "ok": False,
            "errors": [
                {
                    "path": "<stdin>",
                    "code": "invalid_json",
                    "message": f"Failed to parse stdin as JSON: {exc}"
                }
            ]
        }))


def load_inspector_result(payload: Dict[str, Any]) -> Any:
    if "data" in payload:
        return payload["data"]

    path = payload.get("path", "inspector_result.json")
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except FileNotFoundError:
        raise SystemExit(json.dumps({
            "ok": False,
            "errors": [
                {
                    "path": path,
                    "code": "file_not_found",
                    "message": f"inspector result file not found at '{path}'"
                }
            ]
        }))
    except json.JSONDecodeError as exc:
        raise SystemExit(json.dumps({
            "ok": False,
            "errors": [
                {
                    "path": path,
                    "code": "invalid_json",
                    "message": f"inspector result is not valid JSON: {exc}"
                }
            ]
        }))


def validate_inspector_result(data: Any) -> List[ValidationError]:
    errors: List[ValidationError] = []

    if not isinstance(data, dict):
        errors.append(
            ValidationError(
                path="<root>",
                code="type_error",
                message="inspector_result must be a JSON object",
            )
        )
        return errors

    # status: "approved" or "changes_requested"
    status = data.get("status")
    if status not in {"approved", "changes_requested"}:
        errors.append(
            ValidationError(
                path="status",
                code="invalid_enum",
                message="status must be 'approved' or 'changes_requested'",
            )
        )

    # issues: array; non-empty when status === "changes_requested"
    issues = data.get("issues")
    if issues is None:
        errors.append(
            ValidationError(
                path="issues",
                code="required",
                message="issues array is required",
            )
        )
        issues = []
    elif not isinstance(issues, list):
        errors.append(
            ValidationError(
                path="issues",
                code="type_error",
                message="issues must be an array",
            )
        )
        issues = []

    if status == "changes_requested" and not issues:
        errors.append(
            ValidationError(
                path="issues",
                code="empty_for_changes_requested",
                message="issues must be non-empty when status is 'changes_requested'",
            )
        )

    # Validate each issue structure
    for idx, issue in enumerate(issues):
        path_prefix = f"issues[{idx}]"
        if not isinstance(issue, dict):
            errors.append(
                ValidationError(
                    path=path_prefix,
                    code="type_error",
                    message="each issue must be an object",
                )
            )
            continue

        severity = issue.get("severity")
        if severity not in {"blocker", "major", "minor"}:
            errors.append(
                ValidationError(
                    path=f"{path_prefix}.severity",
                    code="invalid_enum",
                    message="severity must be one of 'blocker', 'major', 'minor'",
                )
            )

        description = issue.get("description")
        if not isinstance(description, str) or not description.strip():
            errors.append(
                ValidationError(
                    path=f"{path_prefix}.description",
                    code="required",
                    message="description must be a non-empty string",
                )
            )

        paths = issue.get("paths")
        if not isinstance(paths, list) or not paths:
            errors.append(
                ValidationError(
                    path=f"{path_prefix}.paths",
                    code="required",
                    message="paths must be a non-empty array of strings",
                )
            )
        else:
            for p_idx, p in enumerate(paths):
                if not isinstance(p, str) or not p.strip():
                    errors.append(
                        ValidationError(
                            path=f"{path_prefix}.paths[{p_idx}]",
                            code="type_error",
                            message="each path must be a non-empty string",
                        )
                    )

    # next_tasks: array of strings (may be empty when status is approved)
    next_tasks = data.get("next_tasks")
    if next_tasks is None:
        # Treat missing as empty array (permissive)
        next_tasks = []
    elif not isinstance(next_tasks, list):
        errors.append(
            ValidationError(
                path="next_tasks",
                code="type_error",
                message="next_tasks must be an array of strings",
            )
        )
        next_tasks = []

    for idx, task in enumerate(next_tasks):
        if not isinstance(task, str) or not task.strip():
            errors.append(
                ValidationError(
                    path=f"next_tasks[{idx}]",
                    code="type_error",
                    message="each next_tasks entry must be a non-empty string",
                )
            )

    return errors


def main() -> None:
    payload = load_input()
    result = load_inspector_result(payload)
    errors = validate_inspector_result(result)

    output = {
        "ok": not errors,
        "errors": [e.to_dict() for e in errors],
    }
    sys.stdout.write(json.dumps(output))


if __name__ == "__main__":
    main()
